### PR TITLE
Troubleshooting memory

### PR DESCRIPTION
**Pull request by julianeon**
Heavily edited article and added a memory section. Now starts off with a general advice section, explains in more detail the significance of the 503 error, and talks at length about OOM at bottom.

### CODE REVIEW

File: content/deploy-troubleshooting.md [hunk 1]
```diff
@@ -4,36 +4,51 @@ order: 15
...
 ---
 
-<h2 id="503-errors">503 errors</h2>
+<h2 id="general-advice">General Advice</h2>
 
-The 503 error ("Service Unavailable: No healthy enpoints to handle the request") indicates that our image builder was not able to successfully create a container, when attempting to deploy. This usually indicates a problem in your code that prevents deployment. You may also see this error temporarily, if trying to resolve a URL in your browser within 1-2 minutes of issuing your deploy command.
+Check these items if you're having trouble with uptime, performance or deployment.
+- Your Meteor version. More current versions may resolve issues found in older Meteor versions. While the reverse is rare and generally shouldn't happen, if you're recently upgraded Meteor versions and start having difficulties, consider reverting to your last Meteor version.
+- Your container's memory and CPU usage. If your app is running out of memory, you may need to either switch to a larger container or refactor your app to use less memory. In the short term, the only guaranteed solution is to use a container large enough to handle all your app's memory needs.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: or more containers

File: content/deploy-troubleshooting.md [hunk 2]
```diff
@@ -4,36 +4,51 @@ order: 15
...
+- Your app's logs. While 'All' shows all errors, consider breaking it down by tab. If your app is running and struggling before failing, check the 'App' tab. If your app is crashing before deployment and never successfully launching, check the 'Service' tab. Note that both can happen simultaneously: the earlier version of your app may be throwing errors, while the recent version created to fix the problem may have a code error preventing deployment. In that case, 'Service' will be more helpful.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: "all" shows all output, not just errors.

s/If your app is crashing before deployment and never successfully launching/If your app fails when Galaxy tries to build it into a container image/

File: content/deploy-troubleshooting.md [hunk 3]
```diff
@@ -4,36 +4,51 @@ order: 15
...
+- Check GitHub to see if your issue is listed with a suggested workaround or solution.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: link to http://github.com/meteor/meteor/issues/

File: content/deploy-troubleshooting.md [hunk 4]
```diff
@@ -4,36 +4,51 @@ order: 15
...
+- Consider running more than 1 container, if uptime for your app is mission-critical and you're currently running a single container.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: or can suggest 3 which is "HA"

could be more explicit: "If you only run one container, that means that the machine your container is running on is a single point of failure; in the case of a hardware problem, your app will be down until Galaxy starts it up again on a new machine."

File: content/deploy-troubleshooting.md [hunk 5]
```diff
@@ -4,36 +4,51 @@ order: 15
...
+- Write in to support. To minimize the back-and-forth, please send in the name of the affected app, the conditions that trigger the issue (confirmed or suspected), steps to reproduce, and relevant logs. Try to resolve errors listed in the logs before writing in.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: should "write in to support" be a link?

File: content/deploy-troubleshooting.md [hunk 6]
```diff
@@ -4,36 +4,51 @@ order: 15
...
 
-A common cause of this error is that your `MONGO_URL` variable is missing or is set incorrectly. You can verify what `MONGO_URL` Galaxy is using by going to the app's dashboard and choosing the settings tab. To learn how to set it correctly, check the following resources:
+Note that code-level review lies outside the scope of Galaxy's support.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: link to dev sub?

File: content/deploy-troubleshooting.md [hunk 7]
```diff
@@ -4,36 +4,51 @@ order: 15
...
 
-* [How do I set Environment Variables?](https://galaxy.meteor.com/help/setting-environment-variables) will show you how to set up your `settings.json` file.
+<h2 id="503-errors">503 errors and Deployment failures</h2>
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: a little weird to have D capital and f lowercase

File: content/deploy-troubleshooting.md [hunk 8]
```diff
@@ -4,36 +4,51 @@ order: 15
...
+
+The 503 error will show `Service Unavailable: No healthy endpoints to handle the request` when you try to visit your URL. This indicates that our image builder was not able to successfully create a container, when attempting to deploy. This usually indicates a problem in your code that prevents deployment.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: The first sentence reads strangely.  Maybe more like "If your app shows a 503 error with `Service Unavailable...`,"

Comment (glasser) [inline content/deploy-troubleshooting.md]: Also this isn't really true. This means we are not currently running any healthy containers.  One reason is that all containers are unhealthy — eg they are all stuck in a CPU loop.

Another reason is that none are running — maybe they all just crashed (say, maybe there's only one in the first place and it just crashed and hasn't quite restarted).

Another reason is that a build failed, but this is only really relevant for your initial deploy, or if your previous version is broken for one of the above reasons. We don't start killing old containers until the new version is built and some new containers are running and healthy.

File: content/deploy-troubleshooting.md [hunk 9]
```diff
@@ -4,36 +4,51 @@ order: 15
...
+
+You may see the 503 error temporarily, if trying to resolve a URL in your browser within 1-2 minutes of issuing your deploy command. If you've waited 15 minutes or more, the problem will almost surely not resolve itself by waiting longer.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: This seems to imply that it's expected to get 503s on deploy, which is definitely not true

File: content/deploy-troubleshooting.md [hunk 10]
```diff
@@ -4,36 +4,51 @@ order: 15
...
+
+Begin by checking the logs tab to see if your app is crashing. The 'Service' tab will show build errors. Most of the time, the key to a solution will be found in the exception or error messages. Keep iterating on code fixes and deployments until the error goes away, a new error appears, or your app deploys successfully.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: Somewhat confusingly, the service tab shows both build errors and stuff like "container starting/stopping".  You can tell the difference by seeing if it looks like a container ID (a1b3) or just a version number (v007).

Maybe we should split into more tabs.  The data is marked differently on the backend already.

File: content/deploy-troubleshooting.md [hunk 11]
```diff
@@ -4,36 +4,51 @@ order: 15
...
+
+A common cause of the 503 error is that your `MONGO_URL` variable is missing or is set incorrectly. You can verify what `MONGO_URL` Galaxy is using by going to the app's dashboard and choosing the settings tab. To learn how to set it correctly, check the following resources:
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: Maybe more like "A common reason that your app may be crashing on startup is that"...

File: content/deploy-troubleshooting.md [hunk 12]
```diff
@@ -4,36 +4,51 @@ order: 15
...
+
+* [Environment variables](/environment-variables.html) will show you how to set up your `settings.json` file.
 * [This compose.io article](https://www.compose.io/articles/meteors-new-galaxy-and-the-perfectly-composed-companion/) explains the settings for `MONGO_URL` and `MONGO_OPLOG_URL` in detail.
 
 If you believe your `MONGO_URL` is set correctly, try the following:
 
-* Check the app dashboard and verify the app status is green (healthy)
-* Check the logs tab to see if the app is crashing. If so, look at the exception to determine the issue.
+* Check the app dashboard and verify the app status is green (healthy).
 * Run `dig +show [your app's domain]` in the terminal and verify that its CNAME points to `galaxy-ingress.meteor.com`. You can learn how to set up DNS [here](configuring-dns). If you recently changed your DNS settings, you may need to wait for the new records to propagate. DNS changes often propagate within 30 minutes (depending on the TTL configured for the record set), but in some cases it can take up to 24 hours. Contact your DNS provider if you think there is a problem.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: oops, this is out of date since it's not correct for EU

mention both, and probably include the longer version with us-east-1 in it so it's obvious why there are two.

File: content/deploy-troubleshooting.md [hunk 13]
```diff
@@ -4,36 +4,51 @@ order: 15
...
 
 <h2 id="package-error">Module missing or npm error</h2>
 
-This usually indicates that the module or package working locally in your application is not working after deployment. While third-party packages are unsupported, an explanation of our system may assist you in your troubleshooting.
+This usually indicates that the module or package working locally in your application is not working after deployment. While third-party packages are unsupported, an explanation may help you to troubleshoot.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: this probably should say "While we don't offer support for using specific third-party packages" or something — be clear that it's not that we don't support you at all if you use any 3p packages, just that we won't help you with understanding how to use the.

File: content/deploy-troubleshooting.md [hunk 14]
```diff
@@ -4,36 +4,51 @@ order: 15
...
  
-When you deploy an app, we bundle node_modules into it. npm packages that are required on the client side get built into the bundle uploaded to Galaxy. As a result, Galaxy doesn't need to use npm install for client side bundling to work.
+When you deploy an app, we bundle node_modules into it; npm packages that are required on the client side get built into the bundle uploaded to Galaxy. Galaxy doesn't need to use `npm install` for client side bundling to work.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: maybe s/use/run/

File: content/deploy-troubleshooting.md [hunk 15]
```diff
@@ -46,4 +61,19 @@ Version 1.2.1 of Meteor (and higher) provides the `METEOR_BINARY_DEP_WORKAROUND`
 
 If you are uncertain if this matches your situation, you can use [this test app](https://github.com/zol/meteor-bignum-test) to reproduce the error and confirm a fix.
 
+<h2 id="memory-issues">Memory issues or Out-of-Memory errors</h2>
+
+Memory issues are frequently indicated in your logs. The log message `The container has run out of memory. A new container will be started to replace it.` means that the container running your application tried to get more memory than was allocated to it. Containers in this state are automatically killed.
+
+You can see memory utilization in the container view of your app. Check the 5 minute, 1 hour, 1 day and 30 day views; if you see memory spiking to 90% or more, that is almost certainly the problem.
+
+If your container is running out of memory, you may need to [scale up](/scaling.html) to fix this. You can always scale back down after refactoring your app to consume less memory. If you continue to run your app as is, the only way to prevent it from dying is to allocate containers with more than enough memory.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: again, it's possible that more containers rather than bigger might work too.  depends if it's "need to be able to deal with big spikes occasionally" vs "each connection takes X memory, we have lots of connections".

File: content/deploy-troubleshooting.md [hunk 16]
```diff
@@ -46,4 +61,19 @@ Version 1.2.1 of Meteor (and higher) provides the `METEOR_BINARY_DEP_WORKAROUND`
...
+
+You can also use npm modules to profile your memory usage and pinpoint erratic memory usage. The heapdump npm module is one such module, though you'll need to transfer the file it creates to a place like S3 for download and closer examination. Another such module is memwatch-next.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: links: https://www.npmjs.com/package/heapdump https://www.npmjs.com/package/memwatch-next

File: content/deploy-troubleshooting.md [hunk 17]
```diff
@@ -46,4 +61,19 @@ Version 1.2.1 of Meteor (and higher) provides the `METEOR_BINARY_DEP_WORKAROUND`
...
+
+While Galaxy does not officially support the use of third-party modules, our community has found the use of memory-profiling npm modules to be helpful. Check the GitHub page of your module for more information.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: similar re support/3rd party

File: content/deploy-troubleshooting.md [hunk 18]
```diff
@@ -8,43 +8,53 @@ description: Learn how to troubleshoot your deploy and get answers to frequently
...
+- <a href="mailto:support@meteor.com">Write in</a> to support. To minimize the back-and-forth, please send in the name of the affected app, the conditions that trigger the issue (confirmed or suspected), steps to reproduce, and relevant logs. Try to resolve errors listed in the logs before writing in.
 
-Note that code-level review lies outside the scope of Galaxy's support.
+Note that code-level review lies outside the scope of Galaxy's support. If this is important to you, consider [Meteor Development Support](/support.html).
 
-<h2 id="503-errors">503 errors and Deployment failures</h2>
+<h2 id="503-errors">503 errors and Deployment Failures</h2>
 
-The 503 error will show `Service Unavailable: No healthy endpoints to handle the request` when you try to visit your URL. This indicates that our image builder was not able to successfully create a container, when attempting to deploy. This usually indicates a problem in your code that prevents deployment.
+Your app may throw a 503 error and show `Service Unavailable: No healthy endpoints to handle the request` when you try to visit your URL. This means no healthy containers are currenty available to serve your app.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: "currently" with an L

File: content/deploy-troubleshooting.md [hunk 19]
```diff
@@ -8,43 +8,53 @@ description: Learn how to troubleshoot your deploy and get answers to frequently
...
-- Write in to support. To minimize the back-and-forth, please send in the name of the affected app, the conditions that trigger the issue (confirmed or suspected), steps to reproduce, and relevant logs. Try to resolve errors listed in the logs before writing in.
+- Your container's memory and CPU usage. If your app is running out of memory, you may need to switch to a larger container, use more containers, or refactor your app to use less memory. In the short term, the only guaranteed solution is to use enough Galaxy resources to handle all your app's memory needs.
+- Your app's logs. While 'All' shows all errors, consider breaking it down by tab. If your app is running and struggling before failing, check the 'App' tab. If your app fails when Galaxy tries to build it into a container image, check the 'Service' tab. Note that both can happen simultaneously: the earlier version of your app may be throwing errors, while the recent version created to fix the problem may have a code issue preventing deployment. In that case, 'Service' will be more helpful.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: all shows all output, not just errors

File: content/deploy-troubleshooting.md [hunk 20]
```diff
@@ -8,43 +8,53 @@ description: Learn how to troubleshoot your deploy and get answers to frequently
...
 
-Begin by checking the logs tab to see if your app is crashing. The 'Service' tab will show build errors. Most of the time, the key to a solution will be found in the exception or error messages. Keep iterating on code fixes and deployments until the error goes away, a new error appears, or your app deploys successfully.
+The most common cause for this error is a problem in your code that prevents deployment.
 
-A common cause of the 503 error is that your `MONGO_URL` variable is missing or is set incorrectly. You can verify what `MONGO_URL` Galaxy is using by going to the app's dashboard and choosing the settings tab. To learn how to set it correctly, check the following resources:
+In some cases you may see the 503 error temporarily, if trying to resolve a URL in your browser within 1-2 minutes of issuing your deploy command. If you've waited 15 minutes or more, the problem will almost surely not resolve itself by waiting longer.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: again, it feels like this is saying "Normal Galaxy behavior is to serve 503s for a few minutes after deploying" which is not true.

File: content/deploy-troubleshooting.md [hunk 21]
```diff
@@ -18,13 +18,13 @@ Note that code-level review lies outside the scope of Galaxy's support. If this
...
 
 There are several potential reasons for this. One example is that all containers are unhealthy, because they are all stuck in a CPU loop. Another reason is that none are running, because they all recently crashed (especially if the total number of your containers is 1, and it hasn't had time to restart). Another reason is because your build failed, if this is the first time you're deploying a container for that app or if the only other available containers built successfully but are unhealthy.
 
 The most common cause for this error is a problem in your code that prevents deployment.
 
-In some cases you may see the 503 error temporarily, if trying to resolve a URL in your browser within 1-2 minutes of issuing your deploy command. If you've waited 15 minutes or more, the problem will almost surely not resolve itself by waiting longer.
+In some rare cases you may see the 503 error temporarily, if trying to resolve a URL in your browser within 1-2 minutes of issuing your deploy command. Waiting should resolve this. If you've waited 15 minutes or more, however, the problem will almost surely not resolve itself by waiting longer.
```
Comment (glasser) [inline content/deploy-troubleshooting.md]: I just honestly don't know what you're talking about here.  If "deploys lead to downtime" is a behavior you're hearing about frequently from users, then that's a ginormous Galaxy bug and fixing it should be a high priority.

Comment (julianeon) [inline content/deploy-troubleshooting.md]: I see; I think I've figured out my misunderstanding here. I thought you meant that 503 errors could happen if you checked too quickly, on rare occasions; but I think what you're saying is, there shouldn't be any 503 errors after 1-2 minutes, remove this. I can do that.

The '15 mins or longer' thing comes from users seeing 503 errors and thinking, I'll wait it out; it's really meant to say to them, if you see a 503 error, waiting won't make it go away.

So in summary, it sounds like the whole highlighted green '503 error temporarily' section can be safely removed. If that sounds good to you, let me know and I'll do it.

**Comment by julianeon** [comment]
Removed inaccurate section.
